x/concentrated-liquidity/swaps_test.go

```diff
@@ -1361,3 +1242,85 @@ func (s *KeeperTestSuite) TestSwapExactAmountOut() {
 		})
 	}
 }
+
+func (s *KeeperTestSuite) TestInverseRelationshipSwapOutAmtGivenIn() {
+	tests := swapOutGivenInCases
+
+	for name, test := range tests {
+		s.Run(name, func() {
+			s.Setup()
+			s.FundAcc(s.TestAccs[0], sdk.NewCoins(sdk.NewCoin("eth", sdk.NewInt(10000000000000)), sdk.NewCoin("usdc", sdk.NewInt(1000000000000))))
+			s.FundAcc(s.TestAccs[1], sdk.NewCoins(sdk.NewCoin("eth", sdk.NewInt(10000000000000)), sdk.NewCoin("usdc", sdk.NewInt(1000000000000))))
+
+			// Create default CL pool
+			pool := s.PrepareConcentratedPool()
+
+			// add default position
+			_, _, _, err := s.App.ConcentratedLiquidityKeeper.CreatePosition(s.Ctx, pool.GetId(), s.TestAccs[0], DefaultAmt0, DefaultAmt1, sdk.ZeroInt(), sdk.ZeroInt(), DefaultLowerTick, DefaultUpperTick)
+			s.Require().NoError(err)
+
+			// add second position depending on the test
+			if !test.secondPositionLowerPrice.IsNil() {
+				newLowerTick, err := math.PriceToTick(test.secondPositionLowerPrice, DefaultExponentAtPriceOne) // 315000
+				s.Require().NoError(err)
+				newUpperTick, err := math.PriceToTick(test.secondPositionUpperPrice, DefaultExponentAtPriceOne) // 322500
+				s.Require().NoError(err)
+
+				_, _, _, err = s.App.ConcentratedLiquidityKeeper.CreatePosition(s.Ctx, pool.GetId(), s.TestAccs[1], DefaultAmt0, DefaultAmt1, sdk.ZeroInt(), sdk.ZeroInt(), newLowerTick.Int64(), newUpperTick.Int64())
+				s.Require().NoError(err)
+			}
+
+			// mark pool state and user balance before swap
+			poolBefore, err := s.App.ConcentratedLiquidityKeeper.GetPool(s.Ctx, pool.GetId())
+			s.Require().NoError(err)
+			userBalanceBeforeSwap := s.App.BankKeeper.GetAllBalances(s.Ctx, s.TestAccs[0])
+			poolBalanceBeforeSwap := s.App.BankKeeper.GetAllBalances(s.Ctx, poolBefore.GetAddress())
+
+			// system under test
+			firstTokenIn, firstTokenOut, _, _, _, err := s.App.ConcentratedLiquidityKeeper.SwapOutAmtGivenIn(
+				s.Ctx,
+				test.tokenIn, test.tokenOutDenom,
+				DefaultZeroSwapFee, test.priceLimit, pool.GetId())
+
+			secondTokenIn, secondTokenOut, _, _, _, err := s.App.ConcentratedLiquidityKeeper.SwapOutAmtGivenIn(
+				s.Ctx,
+				firstTokenOut, firstTokenIn.Denom,
+				DefaultZeroSwapFee, sdk.ZeroDec(), pool.GetId(),
+			)
+			s.Require().NoError(err)
+
+			// INVARIANTS
+
+			// 1. assure we get the same tokens after swapping back and forth
+
+			// allow 0.01% of margin of error
+			firstTokenInMargin := firstTokenIn.Amount.Mul(sdk.NewInt(1)).Quo(sdk.NewInt(10000))
+			tokenInDiff := firstTokenIn.Amount.Sub(secondTokenOut.Amount).Abs()
+
+			s.Require().True(firstTokenInMargin.GTE(tokenInDiff))
+			s.Require().Equal(firstTokenOut, secondTokenIn)
```

Comment:
Do you think we can/should use osmomath errTolerance instead? (example from recent test)

https://github.com/osmosis-labs/osmosis/blob/11d6ff0d08d81080ab1d6b3562b398fb7251c399/x/gamm/keeper/msg_server_test.go#L556-L564

Reply:
Nice suggestion! Done!